parla/ldevice.py

```diff
@@ -341,3 +342,58 @@ def wrapper(*args, memory, device):
         def wrapper(*args, memory, device):
             return memory(data(*args))
     return wrapper
+
+
+class Coordinator():
+    def __init__(self, latest_view: List):
+        self._latest_view = latest_view
+
+    def __getitem__(self, index):
+        if not isinstance(index, tuple):
+            index = (index,)
+        ret = []
+
+        def traverse(data, prefix, index):
+            if len(index) > 0:
+                i, *rest = index
+                if isinstance(i, slice):
+                    for v in range(i.start or 0, i.stop, i.step or 1):
+                        traverse(data[v], prefix + (v,), rest)
+                elif isinstance(i, Iterable):
+                    for v in i:
+                        traverse(data[v], prefix + (v,), rest)
+                else:
+                    assert isinstance(i, int)
+                    traverse(data[i], prefix + (i,), rest)
+            else:
+                ret.append(clone_here(data) if is_array(data) else data)
+
+        traverse(self._latest_view, (), index)
```

Comment:
This traversing can be extremely inefficient and I assume it's the major overhead currently.

Reply:
I don't think this will be as much of an overhead as you would think. The traversal creates views at intermediate views not copies. Let's not try to optimize without profiling (which should be easy with `cProfile`). If this is a performance issue, then we can revisit this. Sadly optimizing this will be quite hard since this code uses python dynamic dispatch a lot. One option could be to catch the data views in some form to avoid having to rerun the traversal, but the cache might not be that much faster because of the complexity of the data that would need to be catched.

Reply:
> Let's not try to optimize without profiling (which should be easy with cProfile).
Agreed and will do. Thanks for the suggestions!